teps/add-variable-retries-and-retry-count.md

```diff
@@ -0,0 +1,105 @@
+---
+title: add-variable-retries-and-retry-count
+authors:
+  - "@zhangyumeng"
+creation-date: 2020-10-14
+last-updated: 2020-10-19
+status: proposed
+---
+
+# TEP-0026: Add Variable `retries` and `retry-count`
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Alternatives](#Alternatives)
+- [References](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes adding `retries` and current retry times to Pipeline variables.
+
+## Motivation
+
+PipelineTask can be rerun after failure when user specifies `retries` in PipelineTask. In that case, exposing the retry count can give user more information. User can dispatch a notification after certain times of failure, or customize logs according to retry count.
+
+### Goals
+
+- add 2 variables indicating retries and retry count of a PipelineTask
+
+### Non-Goals
+
+- execute different PipelineTask based on retry count
+- emit an event on every retry
+
+## Proposal
+
+There are 2 variable substitutions `tasks.<taskName>.retries` and `tasks.<taskName>.retry-count` needed to be supported:
+
+1. For retrieving `retries`, we can use `Retries` that stored in `PipelineTask` struct
+2. For retrieving `retry-count`, we can use length of `RetriesStatus` in `TaskRunStatus`
+
+Code example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: failed-task
+spec:
+  params:
+    - name: num-retries
+      type: string
+    - name: max-retries
+      type: string
+  steps:
+    - image: alpine:3.12.0
+      script: |
+        #!/usr/bin/env sh
+        if [ $(params.num-retries) == $(params.max-retries) ]; then
+          echo "This is the last retry. \n"
+        fi
+    - image: ubuntu
+      script: |
+        #!/usr/bin/env sh
+        /path/to/<program-that-fails>
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: retry-example
+spec:
+  tasks:
+  - name: retry-me
+    retries: 5
+    taskRef:
+      name: failed-task
+    params:
+    - name: num-retries
+      value: $(tasks.retry-me.retry-count)
+    - name: max-retries
+      value: $(tasks.retry-me.retries)
+```
+
+In this example, Tekton will echo `This is the last retry` when retry count is equal to 5.
+
+### User Stories
+
+`PipelineTask` author wants to send a Slack notification only if it fails after specified `retries`, not every time it fails.The user wants less interruption.
+
+## Alternatives
+
+Event seems a more decent way to satisfy the user's need. An event should be emitted on a retry. The user could have an event listener (from Triggers) that runs a task to send a notification when the desired event is caught. In this way, Pipelines would not need to expose another variable.
+
+But the drawback is that it would require a lot more setup by the user, such as running Triggers when they might not use it for anything else.
```

Comment:
This is where we need to look at the "mission" of Tekton: "Tekton is a powerful and flexible open-source framework for creating CI/CD systems" (from [tekton.dev](https://tekton.dev/)). 

- What are we optimizing each component (like `tektoncd/pipeline`) for ? 
- Is Tekton meant to provide a full-fledge CI/CD system or just a framework / set of components ?
- If Tekton is meant to provide a full-fledge CI/CD system, should it be in `tektoncd/pipeline` or some opiniated "official" setup (that would be available in the `tektoncd` org) ?

Reply:
This is where "native" notifications in Tekton would help.
I need to create a TEP for this since the old doc is out of date.

Notifications with Tekton are possible today via cloud events + triggers, but they require a complex setup by the user. A "native" notification mechanism would provide a way to simplify the setup for the user.

Reply:
@vdemeester I wonder if it might help at some point to try to draw up some more specific "missions" for each component 🤔  I'm not sure what the answer is for the notifications feature - @afrittoli do you see it being part of pipelines or triggers or a separate project? i could see it either way - but I also could see pipelines + triggers being combined into one project (https://github.com/tektoncd/triggers/issues/697)